rpcs3/util/sysinfo.cpp

```diff
@@ -68,14 +68,76 @@ namespace Darwin_ProcessInfo
 }
 #endif
 
-#if defined(ARCH_ARM64) && defined (_WIN32) && !defined(_MSC_VER)
-static inline u64 __readx18qword(int offset)
+namespace utils
 {
-	u64* peb = nullptr;
-	__asm__ volatile("mov %0, x18" : "=r"(peb));
-	return peb[offset];
-}
+#ifdef _WIN32
+	// Alternative way to read OS version using the registry.
+	static std::string get_fallback_windows_version()
+	{
+		// Some helpers for sanity
+		const auto read_reg_dword = [](HKEY hKey, std::string_view value_name) -> std::pair<bool, DWORD>
+		{
+			DWORD val;
+			DWORD len = sizeof(val);
+			if (ERROR_SUCCESS != RegQueryValueExA(hKey, value_name.data(), nullptr, nullptr, reinterpret_cast<LPBYTE>(&val), &len))
+			{
+				return { false, 0 };
+			}
+			return { true, val };
+		};
+
+		const auto read_reg_sz = [](HKEY hKey, std::string_view value_name) -> std::pair<bool, std::string>
+		{
+			char sz[256];
+			DWORD sz_len = sizeof(sz);
+			if (ERROR_SUCCESS != RegQueryValueExA(hKey, value_name.data(), nullptr, nullptr, reinterpret_cast<LPBYTE>(sz), &sz_len))
+			{
+				return { false, "" };
+			}
+
+			if (sz_len < sizeof(sz))
+			{
+				// Safety, force null terminator
+				sz[sz_len] = 0;
+			}
+			return { true, sz };
+		};
+
+		HKEY hKey;
+		if (ERROR_SUCCESS != RegOpenKeyExA(HKEY_LOCAL_MACHINE, "SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion", 0, KEY_READ, &hKey))
+		{
+			return "Unknown Windows";
+		}
+
+		// Read ProductName (string), CurrentMajorVersionNumber (dword), CurrentMinorVersionNumber (dword) and CurrentBuildNumber (string) as well as CurrentVersion (string)
+		const auto [product_valid, product_name] = read_reg_sz(hKey, "ProductName");
+		if (!product_valid)
+		{
+			return "Unknown Windows";
+		}
+
+		const auto [check_major, version_major] = read_reg_dword(hKey, "CurrentMajorVersionNumber");
+		const auto [check_minor, version_minor] = read_reg_dword(hKey, "CurrentMinorVersionNumber");
+		const auto [check_build_no, build_no] = read_reg_sz(hKey, "CurrentBuildNumber");
+		const auto [check_nt_ver, nt_ver] = read_reg_sz(hKey, "CurrentVersion");
+
+		// Close the registry key
+		RegCloseKey(hKey);
```

Comment:
Missing close above

Reply:
What?

Reply:
before the previous "unknown windows" return.

Reply:
Right, thanks.